tensorboard/uploader/uploader_main.py

```diff
@@ -502,12 +502,10 @@ def _get_intent(flags):
 
 
 def _get_server_info(flags):
-  if flags.api_endpoint and not flags.origin:
-    return server_info_lib.create_server_info(
-        "http://localhost:8080", flags.api_endpoint)
   origin = flags.origin or _DEFAULT_ORIGIN
-  server_info = server_info_lib.fetch_server_info(
-      flags.origin or _DEFAULT_ORIGIN)
+  if flags.api_endpoint and not flags.origin:
```

Comment:
I think we actually should also remove the argparse-facing default value for `--api_endpoint` too, because otherwise passing `--origin=https://staging.tensorboard.dev` will not behave as desired, since the `flags.api_endpoint` override behavior will still kick in and default to the prod API.  We can instead just set the hardcoded API endpoint as a fallback behavior used for the `create_server_info()` path only, which gets us closer to our desired final state.

Then we'd have the following code for now:

```python
_DEFAULT_ORIGIN = 'https://tensorboard.dev'
_HARDCODED_API_ENDPOINT = 'api.tensorboard.dev:443'

def _get_server_info(flags):
  origin = flags.origin or _DEFAULT_ORIGIN
  if not flags.origin:
    # Temporary fallback to hardcoded API endpoint in the unspecified case.
    api_endpoint = flags.api_endpoint or _HARDCODED_API_ENDPOINT
    return server_info_lib.create_server_info(origin, api_endpoint)
  # same from here down
```

And after the server-side rollout, we'd remove `_HARDCODED_API_ENDPOINT` and update it to:

```python
if not flags.origin and flags.api_endpoint:
  return server_info_lib.create_server_info(origin, flags.api_endpoint)
```

Reply:
> I think we actually should also remove the argparse-facing default
> value for `--api_endpoint` too, because otherwise passing
> `--origin=https://staging.tensorboard.dev` will not behave as desired,
> since the `flags.api_endpoint` override behavior will still kick in
> and default to the prod API

Correct (this is why the test plan passes `--api_endpoint ""` explicitly
in each case). This was fine with me, as it’s only temporary and only
for devs, so I preferred to keep the minor inconvenience so that the
later change could just be a simple flag flip rather than a flag flip
plus removing some logic.

Does this make sense? I’m happy to make the change if you still prefer
it that way.

Reply:
Ok, sorry, should have also read the updated test plan.

That said, even though it's only dev-facing, I don't really love that it's possible to run `--origin=https://staging.tensorboard.dev` and pretty misleading behavior (data uploaded to the wrong place, printed URL doesn't resolve).  It seems pretty easy for one of us to accidentally hit that during testing.  I'd be more comfortable making that work correctly now; it seems fine to me that the later change removes fallback logic since I think that's actually more reflective of what we're doing than just flipping a flag.

Reply:
No problem; that makes sense. Done. I think that you’re probably right,
as this simplifies the test plan, which is a good sign. :-)